### ISSUE TITLE
Why use "unfold" to transfer shape of token_logits from (B, C, H, W) to (B, C, 9, H, W) ?

### ISSUE THREAD

**Issue by whiteinblue** [YiF-Zhang/RegionProxy#3]

**Comment by YiF-Zhang** [comment]
Hi @whiteinblue , it's an efficient implementation of the "painting" process described in Eq.6. For each output pixel, we are actually aggregating the predicted token logits of the regions around it. The process is actually similar to the convolution, i.e., im2col followed by matrix multiplication. We first acquire the logits of nearby regions, then aggregate them using the predicted affinity. The "unfold" code do the first step.

 Please refer to the discription in Section 4.3 for better understanding. Feel free to leave comments if you have further questions.

**Comment by whiteinblue** [comment]
Thanks for your reply,  the region you explained is a regular 3*3 grid  patch feature from spatial token features, which correspond a h*w pixel patch. Look backward, every pixel  in h*w patch,  you use Q^T  to aggregate pixel info to generate regular 3*3 spatial token features. So the implict regions is not irregular regions like superpixel, just a 3*3 patch. understand right?

**Comment by YiF-Zhang** [comment]
That's not correct. The affinity map its self describe the geometrics of the regions, it's not about aggregating pixel information and it does not contain semantics (it's class-agnostic). For your "implicit regions" comment, the tokens do not imply region geometrics, so there is no "implicit regions". In short, the "region" includes two parts: its geometrics described by affinity, and its feature encoded as tokens.

For the last sentence of your comment, note the region is concrete. Although it's soft, it can be quantized and visualized (see Figure.6 & Figure.8).

**Comment by whiteinblue** [comment]
I see, Thanks
